File: macos/Sources/SnapikCore/Geometry/StripResizeGeometry.swift
// Port of `src/Snapik.App/Controls/StripResizeGeometry.cs`, SPEC-DELTA-3 §1.3 S-9, §2.5,
// SPEC-DELTA-4 §2.5, §3.1.
import Foundation

/// The width of the strip, dragged by its left edge. The right edge is fixed while the drag lasts,
/// so the card grows into the screen instead of walking away from it. Sizes are window sizes: the
/// visible panel is 20 points narrower on each side, and that field carries the shadow
/// (`StackMetrics.shadowMargin`, the Stack zone).
///
/// Every figure is taken **from the start of the drag**, never as a sum of deltas: a delta added to
/// the width of the moment keeps the travel already spent beyond a clamp, and the strip then ignores
/// the whole way back until the pointer has given that travel up again — the dead zone the strip was
/// reported to have.
public enum StripResizeGeometry {
    /// 244 on both edges: the panel the user sees is 204 wide and the field under the shadow takes
    /// 20 on each side (`204 + 2 · 20`). A width stored by a build whose floor was 200, 208 or the
    /// 224 of the intermediate round is lifted to this one by `clampWidth`: the panel it stood for
    /// is narrower than the one the strip draws now (SPEC-DELTA-4 §2.5).
    public static let minimumWidth: Double = 244
    public static let defaultWidth: Double = 244

    /// The gap the strip keeps between itself and the right edge of the working area. It is also the
    /// ceiling of the width: a strip as wide as the whole working area would have to start outside
    /// of it to keep that gap. The field under the shadow is the gap that is seen now, so this one
    /// is zero and the panel still stands 20 points away from the edge.
    public static let edgeGap: Double = 0

    /// The height of the strip is the height of the capture list, not of the window: the window
    /// derives its own height from this one. The card and the overlap stay as they are, the visible
    /// part of the list is what grows. There is no number above: the working area of the monitor is
    /// the only ceiling.
    public static let minimumListHeight: Double = 180
    public static let defaultListHeight: Double = 372

    /// Everything of the strip window that is not the list: the header, the capture button, the
    /// toast, the status line and the paddings of the card. The window measures its own before the
    /// first drag; this is the figure used until there is something to measure, and it is on the
    /// generous side on purpose — a list clamped a little short still fits on the screen.
    public static let estimatedChromeHeight: Double = 160

    /// A width from the settings file. The ceiling is the working area of the screen the strip opens
    /// on, less the gap it keeps at the edge, so a width dragged out on a large monitor is pulled
    /// back in on a small one; nonsense falls back to the default. When the working area is unknown
    /// there is no ceiling, only the minimum.
    public static func clampWidth(_ width: Double, workWidth: Double) -> Double {
        let ceiling =
            workWidth.isFinite && workWidth > 0
            ? max(minimumWidth, workWidth - edgeGap)
            : Double.infinity
        if !width.isFinite { return min(defaultWidth, ceiling) }
        return min(max(width, minimumWidth), ceiling)
    }

    /// The left edge and the width for a pointer that has travelled `pointerDelta` points from where
    /// it was when the drag began, measured against `startWidth`, the width the strip had at that
    /// same moment. The strip never crosses `leftLimit`, the left edge of the working area it sits in.
    public static func widthFromStart(
        right: Double,
        startWidth: Double,
        pointerDelta: Double,
        leftLimit: Double
    ) -> (left: Double, width: Double) {
        let start = startWidth.isFinite ? startWidth : defaultWidth
        let delta = pointerDelta.isFinite ? pointerDelta : 0
        let allowed = max(minimumWidth, right - leftLimit)
        let resized = min(max(start - delta, minimumWidth), allowed)
        return (right - resized, resized)
    }

    /// A list height from the settings file: out of range is clamped, nonsense falls back to the
    /// default. The working area is a ceiling of its own, and the window is taller than its list by
    /// `chromeHeight`, so a height stored on a large monitor opens a window that still fits on the
    /// screen it comes back on, together with the corner grip that resizes it.
    public static func clampListHeight(_ height: Double, workHeight: Double, chromeHeight: Double)
        -> Double
    {
        let chrome = chromeHeight.isFinite && chromeHeight > 0 ? chromeHeight : estimatedChromeHeight
        let ceiling =
            workHeight.isFinite && workHeight > 0
            ? max(minimumListHeight, workHeight - chrome)
            : Double.infinity
        let stored = height.isFinite ? height : defaultListHeight
        return min(max(stored, minimumListHeight), ceiling)
    }

    /// The height of the capture list for a pointer that has travelled `pointerDelta` points
    /// downwards from where it was when the drag began, measured against `startListHeight`, the
    /// height the list had at that same moment — the way back from a clamp then moves the strip on
    /// the first pixel. The top edge stays where it was, so the strip grows down; `chromeHeight` is
    /// everything of the window that is not the list, and it keeps the bottom of the window inside
    /// `workBottom`.
    public static func listHeightFromStart(
        startListHeight: Double,
        pointerDelta: Double,
        chromeHeight: Double,
        top: Double,
        workBottom: Double
    ) -> Double {
        let start = startListHeight.isFinite ? startListHeight : defaultListHeight
        let delta = pointerDelta.isFinite ? pointerDelta : 0
        let available = workBottom - top - max(0, chromeHeight)
        let ceiling = max(minimumListHeight, available)
        return min(max(start + delta, minimumListHeight), ceiling)
    }

    // MARK: - The list by its contents (SPEC-DELTA-5 §1.10, `StripResizeGeometry.cs:37-103`)

    /// The hint that stands where the list is while the strip holds nothing.
    public static let emptyListHeight: Double = 92

    public static let listTopPadding: Double = 14
    public static let listBottomPadding: Double = 8

    /// The paddings of the capture list are not the same on both sides: the right one carries the
    /// scroll bar, which stands over the cards and needs a field of its own, and the left one gives
    /// those four points back so that the card keeps the 168 of the reference shot. The strip reads
    /// them through `StackMetrics`, which derives its numbers from these ones.
    public static let listPaddingLeft: Double = 4
    public static let listPaddingRight: Double = 12

    /// A card and the part of it the next card lies over; the difference is the pitch.
    public static let cardHeight: Double = 78
    public static let cardOverlap: Double = 48
    public static let cardPitch: Double = cardHeight - cardOverlap

    /// The height the capture list wants for `count` cards, and the ceiling the corner grip has
    /// written into the settings. The grip sets the ceiling, not the height: a list of two cards is
    /// 130 tall whatever the settings say, and it stops growing at the ceiling.
    /// `minimumListHeight` is no floor here — it belongs to the stored number alone, otherwise a
    /// single capture would open a list of 180 instead of 100.
    public static func listHeightForCount(_ count: Int, cap: Double) -> Double {
        if count <= 0 { return emptyListHeight }
        let content = listTopPadding + Double(count - 1) * cardPitch + cardHeight + listBottomPadding
        let ceiling = cap.isFinite && cap > 0 ? cap : defaultListHeight
        return min(content, ceiling)
    }

    /// The height of the list: dragged by hand it is the stored number itself, automatic it is the
    /// height of what the list holds. The stored number and the clamps it went through are the same
    /// in both; what the drag changes is whether that number is a ceiling or the height.
    public static func listHeight(count: Int, stored: Double, manual: Bool) -> Double {
        manual ? stored : listHeightForCount(count, cap: stored)
    }

    /// The left edge of the capsule, so that it keeps the right edge of the strip it came from
    /// rather than the edge of the monitor: both windows carry the same field under their shadow,
    /// so the sides the user sees line up.
    public static func capsuleLeft(stripLeft: Double, stripWidth: Double, capsuleWidth: Double)
        -> Double
    {
        stripLeft + stripWidth - capsuleWidth
    }

    /// The rectangle the strip had before the capsule, moved back into the working area only when it
    /// no longer fits: a strip dragged away from the edge stays where the user left it.
    ///
    /// Plain numbers and not a `CGRect`, because Core knows Foundation alone (SPEC-DELTA-5 §2.5);
    /// the caller in the Stack zone takes the two rectangles apart. The width and the height never
    /// move, so the answer is the point. `work` is read as "from `workX`/`workY`, `workWidth` wide
    /// and `workHeight` tall", which is the same pair of edges on either axis direction: what
    /// Windows calls `Top`/`Bottom` is the smaller and the larger edge here too.
    public static func restoreRect(
        x: Double, y: Double, width: Double, height: Double,
        workX: Double, workY: Double, workWidth: Double, workHeight: Double
    ) -> (x: Double, y: Double) {
        let workRight = workX + workWidth
        let workBottom = workY + workHeight
        // The order of the guards is the one of `!work.IsEmpty && work.Width > 0 &&
        // work.Height > 0 && !work.Contains(stored)`: an empty working area is exactly the one with
        // no width or no height here, so the first of the four guards has nothing of its own left
        // to ask. `Rect.Contains(Rect)` is "every edge of the stored one is inside".
        let contains =
            workX <= x && workY <= y && workRight >= x + width && workBottom >= y + height
        if workWidth > 0 && workHeight > 0 && !contains {
            return (
                min(max(x, workX), max(workX, workRight - width)),
                min(max(y, workY), max(workY, workBottom - height))
            )
        }
        return (x, y)
    }
}

File: src/Snapik.App/Controls/StripResizeGeometry.cs
using System;
using System.Windows;

namespace Snapik.App.Controls;

/// <summary>
/// The width of the strip, dragged by its left edge. The right edge is fixed while the drag lasts,
/// so the card grows into the screen instead of walking away from it. Sizes are window sizes: the
/// visible card is 20 px narrower, that margin carries the shadow.
/// </summary>
internal static class StripResizeGeometry
{
    /// <summary>
    /// The transparent field around the panel that carries the shadow of the window: a blur radius
    /// of 24 and a depth of 5 need 12 + 5 = 17 px, and 20 is the round number above that. It is the
    /// margin of the shell, of the capsule and of the grips.
    /// </summary>
    internal const double ShadowMargin = 20;

    /// <summary>The panel the user sees: the window without the field under its shadow.</summary>
    internal const double MinimumPanelWidth = 204;

    internal const double MinimumWidth = MinimumPanelWidth + 2 * ShadowMargin;
    internal const double DefaultWidth = MinimumWidth;

    /// <summary>
    /// The gap the strip keeps between itself and the right edge of the working area. It is also
    /// the ceiling of the width: a strip as wide as the whole working area would have to start
    /// outside of it to keep that gap. The field under the shadow is the gap that is seen now, so
    /// this one is zero and the panel still stands 20 px away from the edge.
    /// </summary>
    internal const double EdgeGap = 0;

    /// <summary>The two paddings between the panel and a card: the shell and the capture list.</summary>
    internal const double ShellPadding = 10;

    /// <summary>
    /// The paddings of the capture list are not the same on both sides: the right one carries the
    /// scroll bar, which stands over the cards and needs a field of its own, and the left one gives
    /// the four pixels back so that the card keeps the 168 px of the reference shot.
    /// </summary>
    internal const double ListPaddingLeft = 4;
    internal const double ListPaddingRight = 12;

    /// <summary>The card of a capture inside a window of that width: 168 px at 244.</summary>
    internal static double CardWidth(double windowWidth) =>
        windowWidth - 2 * ShadowMargin - 2 * ShellPadding - ListPaddingLeft - ListPaddingRight;

    /// <summary>The hint that stands where the list is while the strip holds nothing.</summary>
    internal const double EmptyListHeight = 92;

    internal const double ListTopPadding = 14;
    internal const double ListBottomPadding = 8;

    /// <summary>A card and the part of it the next card lies over; the difference is the pitch.</summary>
    internal const double CardHeight = 78;
    internal const double CardOverlap = 48;
    internal const double CardPitch = CardHeight - CardOverlap;

    /// <summary>
    /// The height the capture list wants for <paramref name="count"/> cards, and the ceiling the
    /// corner grip has written into the settings. The grip sets the ceiling, not the height: a list
    /// of two cards is 130 tall whatever the settings say, and it stops growing at the ceiling.
    /// <see cref="MinimumListHeight"/> is no floor here — it belongs to the stored number alone,
    /// otherwise a single capture would open a list of 180 instead of 100.
    /// </summary>
    internal static double ListHeightForCount(int count, double cap)
    {
        if (count <= 0) return EmptyListHeight;
        var content = ListTopPadding + (count - 1) * CardPitch + CardHeight + ListBottomPadding;
        var ceiling = double.IsFinite(cap) && cap > 0 ? cap : DefaultListHeight;
        return Math.Min(content, ceiling);
    }

    /// <summary>
    /// The left edge of the capsule, so that it keeps the right edge of the strip it came from
    /// rather than the edge of the monitor: both windows carry the same field under their shadow,
    /// so the sides the user sees line up.
    /// </summary>
    internal static double CapsuleLeft(double stripLeft, double stripWidth, double capsuleWidth) =>
        stripLeft + stripWidth - capsuleWidth;

    /// <summary>
    /// The rectangle the strip had before the capsule, moved back into the working area only when it
    /// no longer fits: a strip dragged away from the edge stays where the user left it.
    /// </summary>
    internal static Rect RestoreRect(Rect stored, Rect work)
    {
        if (!work.IsEmpty && work.Width > 0 && work.Height > 0 && !work.Contains(stored))
        {
            var x = Math.Min(Math.Max(stored.X, work.Left), Math.Max(work.Left, work.Right - stored.Width));
            var y = Math.Min(Math.Max(stored.Y, work.Top), Math.Max(work.Top, work.Bottom - stored.Height));
            return new Rect(x, y, stored.Width, stored.Height);
        }
        return stored;
    }

    // The height of the strip is the height of the capture list: the window lives on
    // SizeToContent="Height", and a height written to the window itself is overwritten by the next
    // layout pass. The card (78 px) and the overlap (48 px) stay as they are, the visible part of
    // the list is what grows. The list carries that height outright rather than as a maximum: with
    // a maximum a short strip is shorter than the number it holds, the window stops following the
    // corner grip, and the drag moves nothing until the tenth capture. There is no number above:
    // the working area of the monitor is the only ceiling.
    internal const double MinimumListHeight = 180;
    internal const double DefaultListHeight = 372;

    /// <summary>
    /// Everything of the strip window that is not the list: the header, the capture button, the
    /// toast, the status line and the paddings of the card. The window measures its own before the
    /// first drag; this is the figure used until there is something to measure, and it is on the
    /// generous side on purpose — a list clamped a little short still fits on the screen.
    /// </summary>
    internal const double EstimatedChromeHeight = 160;

    /// <summary>
    /// A width from the settings file. The ceiling is the working area of the monitor the strip
    /// opens on, less the gap it keeps at the edge, so a width dragged out on a large monitor is
    /// pulled back in on a small one; nonsense falls back to the default. When the working area is
    /// unknown there is no ceiling, only the minimum.
    /// </summary>
    internal static double ClampWidth(double width, double workWidth)
    {
        var ceiling = double.IsFinite(workWidth) && workWidth > 0
            ? Math.Max(MinimumWidth, workWidth - EdgeGap)
            : double.PositiveInfinity;
        if (!double.IsFinite(width)) return Math.Min(DefaultWidth, ceiling);
        return Math.Clamp(width, MinimumWidth, ceiling);
    }

    /// <summary>
    /// The left edge and the width for a pointer that has travelled <paramref name="pointerDelta"/>
    /// pixels from where it was when the drag began, measured against <paramref name="startWidth"/>,
    /// the width the strip had at that same moment. Both figures come from the start of the drag on
    /// purpose: a delta added to the width of the moment keeps the travel already spent beyond a
    /// clamp, and the strip then ignores the whole way back until the pointer has given that travel
    /// up again. The strip never crosses <paramref name="leftLimit"/>, the left edge of the working
    /// area it sits in.
    /// </summary>
    internal static (double Left, double Width) WidthFromStart(double right, double startWidth, double pointerDelta, double leftLimit)
    {
        if (!double.IsFinite(startWidth)) startWidth = DefaultWidth;
        if (!double.IsFinite(pointerDelta)) pointerDelta = 0;
        var allowed = Math.Max(MinimumWidth, right - leftLimit);
        var resized = Math.Clamp(startWidth - pointerDelta, MinimumWidth, allowed);
        return (right - resized, resized);
    }

    /// <summary>
    /// A list height from the settings file: out of range is clamped, nonsense falls back to the
    /// default. What comes out is the ceiling the list grows to, not the height it is given — that
    /// one is <see cref="ListHeightForCount"/>, and the stored number only stops it.
    /// The working area is a ceiling of its own, and the window is taller than its list by
    /// <paramref name="chromeHeight"/>, so a height stored on a large monitor opens a window that
    /// still fits on the screen it comes back on, together with the corner grip that resizes it.
    /// </summary>
    internal static double ClampListHeight(double height, double workHeight, double chromeHeight)
    {
        var chrome = double.IsFinite(chromeHeight) && chromeHeight > 0 ? chromeHeight : EstimatedChromeHeight;
        var ceiling = double.IsFinite(workHeight) && workHeight > 0
            ? Math.Max(MinimumListHeight, workHeight - chrome)
            : double.PositiveInfinity;
        return Math.Clamp(double.IsFinite(height) ? height : DefaultListHeight, MinimumListHeight, ceiling);
    }

    /// <summary>
    /// The height of the capture list for a pointer that has travelled <paramref name="pointerDelta"/>
    /// pixels downwards from where it was when the drag began, measured against
    /// <paramref name="startListHeight"/>, the height the list had at that same moment — the way back
    /// from a clamp then moves the strip on the first pixel. The top edge stays where it was, so the
    /// strip grows down; <paramref name="chromeHeight"/> is everything of the window that is not the
    /// list (header, button, toast, paddings), and it keeps the bottom of the window inside
    /// <paramref name="workBottom"/>.
    /// </summary>
    internal static double ListHeightFromStart(double startListHeight, double pointerDelta, double chromeHeight, double top, double workBottom)
    {
        if (!double.IsFinite(startListHeight)) startListHeight = DefaultListHeight;
        if (!double.IsFinite(pointerDelta)) pointerDelta = 0;
        var available = workBottom - top - Math.Max(0, chromeHeight);
        var ceiling = Math.Max(MinimumListHeight, available);
        return Math.Clamp(startListHeight + pointerDelta, MinimumListHeight, ceiling);
    }

    /// <summary>
    /// A working area read from the monitor, in pixels, in the units the window is placed in. The
    /// strip may live on a second monitor with a scale of its own, and the placement and the drag
    /// limit must both come from that one, not from the primary screen.
    /// </summary>
    internal static Rect ToDeviceIndependent(Rect area, double dpiScaleX, double dpiScaleY)
    {
        var scaleX = dpiScaleX > 0 ? dpiScaleX : 1;
        var scaleY = dpiScaleY > 0 ? dpiScaleY : 1;
        return new Rect(area.X / scaleX, area.Y / scaleY, area.Width / scaleX, area.Height / scaleY);
    }
}
